pkg/jobs/job_info_storage.go

```diff
@@ -13,18 +13,174 @@ package jobs
+		if !ok {
+			return 0, hlc.Timestamp{}, time.Time{}, errors.AssertionFailedf("job progress: expected value to be DFloat (was %T)", row[1])
+		}
+		fraction = float64(*fracDatum)
+	}
+
+	var ts hlc.Timestamp
+	if row[2] != tree.DNull {
+		resolved, ok := row[2].(*tree.DDecimal)
+		if !ok {
+			return 0, hlc.Timestamp{}, time.Time{}, errors.AssertionFailedf("job progress: expected value to be DDecimal (was %T)", row[1])
+		}
+		ts, err = hlc.DecimalToHLC(&resolved.Decimal)
+		if err != nil {
+			return 0, hlc.Timestamp{}, time.Time{}, err
+		}
+	}
+
+	return fraction, ts, written.Time, nil
+}
+
+// Write writes a progress update. If fraction is NaN or resolved is empty, that
+// field is left null. The time at which the progress was reported is recorded.
+func (i ProgressStorage) Write(ctx context.Context, txn isql.Txn, fraction float64, resolved hlc.Timestamp) error {
+	ctx, sp := tracing.ChildSpan(ctx, "write-job-progress")
+	defer sp.Finish()
+
+	if _, err := txn.ExecEx(
+		ctx, "write-job-progress-delete", txn.KV(),
+		sessiondata.NodeUserSessionDataOverride,
+		`DELETE FROM system.job_progress WHERE job_id = $1`, i,
+	); err != nil {
+		return err
+	}
+
+	var frac, ts interface{}
+	if !math.IsNaN(fraction) {
+		frac = fraction
+	}
+	if resolved.IsSet() {
+		ts = resolved.AsOfSystemTime()
+	}
+	_, err := txn.ExecEx(
+		ctx, "write-job-progress-insert", txn.KV(),
+		sessiondata.NodeUserSessionDataOverride,
+		`INSERT INTO system.job_progress (job_id, written, fraction, resolved) VALUES ($1, now(), $2, $3)`,
+		i, frac, ts,
+	)
+	return err
+}
+
+type JobStatusKind string
+
+const (
+	JobStatusInfo  JobStatusKind = "info" // sometimes called "running status"
+	JobStatusTrace JobStatusKind = "trace"
+	JobStatusState JobStatusKind = "state"
+)
+
+// StatusStorage reads and writes status rows.
+type StatusStorage jobspb.JobID
+
+// StatusStorage returns a new StatusStorage with the passed in job and txn.
+func (j *Job) StatusStorage() StatusStorage {
+	return StatusStorage(j.id)
+}
+
+// Write writes a status message of the specified kind.
+func (i StatusStorage) Write(ctx context.Context, txn isql.Txn, kind JobStatusKind, message string) error {
+	ctx, sp := tracing.ChildSpan(ctx, "write-job-status")
+	defer sp.Finish()
+
+	_, err := txn.ExecEx(
+		ctx, "write-job-status-insert", txn.KV(),
+		sessiondata.NodeUserSessionDataOverride,
+		`INSERT INTO system.job_status (job_id, kind, written, message) VALUES ($1, $2, now(), $3)`,
+		i, kind, message,
+	)
+	return err
+}
+
+func (i StatusStorage) Latest(ctx context.Context, txn isql.Txn, kindFilter JobStatusKind) (string, JobStatusKind, time.Time, error) {
+	ctx, sp := tracing.ChildSpan(ctx, "get-job-status")
+	defer sp.Finish()
+
+	row, err := txn.QueryRowEx(
+		ctx, "job-status-get", txn.KV(),
+		sessiondata.NodeUserSessionDataOverride,
+		"SELECT written, kind, message FROM system.job_status WHERE job_id = $1 AND (kind = $2 OR $2 = '') ORDER BY written DESC LIMIT 1",
```

Comment:
what's the deal with this `(kind = $2 OR $2 = '')` shenanigans?

Reply:
if you pass an empty kind filter you just get the most recent status of any kind.

Reply:
I see. To make this clearer, I think you should define an empty ` JobStatusKind ` const above and document its semantics.